optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/chained/ChainedSwapMove.java

```diff
@@ -31,26 +32,28 @@
  */
 public class ChainedSwapMove<Solution_> extends SwapMove<Solution_> {
 
-    protected final List<Object> oldLeftTrailingEntityList;
-    protected final List<Object> oldRightTrailingEntityList;
+    protected final Map<GenuineVariableDescriptor<Solution_>, Object> oldLeftTrailingEntityMap;
+    protected final Map<GenuineVariableDescriptor<Solution_>, Object> oldRightTrailingEntityMap;
 
     public ChainedSwapMove(List<GenuineVariableDescriptor<Solution_>> variableDescriptorList,
-            List<SingletonInverseVariableSupply> inverseVariableSupplyList, Object leftEntity, Object rightEntity) {
+            Map<GenuineVariableDescriptor<Solution_>, SingletonInverseVariableSupply> inverseVariableSupplyMap,
+            Object leftEntity, Object rightEntity) {
         super(variableDescriptorList, leftEntity, rightEntity);
-        oldLeftTrailingEntityList = new ArrayList<>(inverseVariableSupplyList.size());
-        oldRightTrailingEntityList = new ArrayList<>(inverseVariableSupplyList.size());
-        for (SingletonInverseVariableSupply inverseVariableSupply : inverseVariableSupplyList) {
-            oldLeftTrailingEntityList.add(inverseVariableSupply.getInverseSingleton(leftEntity));
-            oldRightTrailingEntityList.add(inverseVariableSupply.getInverseSingleton(rightEntity));
```

Comment:
This is where the null would be thrown.
I fix that by not iterating over the cases where the null would be thrown. (Non-chained vars.)

Reply:
I'd iterate over them, to avoid changing an indexed lookup by a hashed lookup later on, but put in null if they don't exist, instead of throwing an NPE. wdyt?